src/twisted/protocols/_smb/_base.py

```diff
@@ -0,0 +1,396 @@
+    if remainder <= ERROR:
+        return obj
+    elif remainder == OFFSET:
+        return (obj, offset + strct.size)
+    else:
+        return (obj, data[offset + strct.size :])
+
+
+def smb_fields(cls: Any) -> List[Any]:
+    return [i for i in attr.fields(cls) if SMB_METADATA in i.metadata]
+
+
+def _get_struct(cls: Any) -> struct.Struct:
+    try:
+        # we use classes to hold cache of Structs as precompiling is more
+        # efficient
+        strct = cast(struct.Struct, cls._struct)
+    except AttributeError:
+        strct = struct.Struct(
+            "<" + "".join(i.metadata[SMB_METADATA] for i in smb_fields(cls))
+        )
+        cls._struct = strct
+    return strct
+
+
+def calcsize(cls: Any) -> int:
+    """
+    return the size of a structure.
+    @param cls: C{attr} decorated class
+    @rtype: L{int}
+    """
+    strct = _get_struct(cls)
+    return strct.size
+
+
+_leint = struct.Struct("<I")
+
+
+def int32key(d: Dict[int, Any], val: Any) -> int:
+    """
+    generate a new random key for a dictionary
+    @param d: dictionary with 32-bit int keys
+    @type d: L{dict}
+    @param val: new dictionary value
+    @rtype: L{int}
+    """
+    assert len(d) < 0xC0000000  # otherwiae dict so big hard to find keys
+    n = 0
+    while n == 0 or n in d:
+        [n] = _leint.unpack(secureRandom(_leint.size, True))
+    d[n] = val
+    return n
+
+
+@attr.s
+class SMBPacket:
+    """
+    A SMB packet as it moves through the system.
+
+    @ivar data: raw data of the packet, both reception and
+                transmission.
+    @type data: L{bytes}
+
+    @ivar hdr: the parsed header
+    @ivar body: the parsed body
+    """
+
+    data: bytes = attr.ib()
+    _proto: "SMBPacketReceiver" = attr.ib()
+    hdr: Any = attr.ib(default=None)
+    body: Any = attr.ib(default=None)
+
+    @property
+    def ctx(self) -> Dict[str, Any]:
+        """
+        the connection context: objects that need to persist scross
+        packets
+        @rtype: L{dict}
+        """
+        return self._proto.ctx
+
+    def send(self) -> None:
+        """
+        transmit the packet's data
+        """
+        self._proto.sendPacket(self.data)
+
+    def close(self) -> None:
+        """
+        close the underlying connection
+        """
+        cast(ITransport, self._proto.transport).loseConnection()
+
+    def clone(self, **kwargs: Any) -> "SMBPacket":
+        """
+        a new packet associated with the same connection
+        @rtype: L{SMBPacket}
+        """
+        kwargs["proto"] = self._proto
+        return SMBPacket(**kwargs)
+
+
+BASE_HEADER = struct.Struct("!xBH")
+
+
+class SMBPacketReceiver(protocol.Protocol):
+    """
+    A L{SMBPacketReceiver} is a wire protocol parser for the SMB framing
+    mechanism, which consist of a 4-byte header: single null
+    and a 24-bit length field.
+    """
+
+    def __init__(
+        self, packetReceived: Callable[[SMBPacket], None], ctx: Dict[str, Any]
```

Comment:
Love having `packetReceived` available via composition rather than needing to literally inherit :)